### PR TITLE
Preserve aliases information from merged packages

### LINKED ISSUE #100: aliases on requires from included configs don't seem to work
I have a composer.json with a requirement on package A and a requirement on package B:

```
{
  "require": {
    "A": "dev-trunk",
    "B": "dev-trunk"
  }
}
```

package B also has a requirement on package `A @ dev-trunk`.

If I try to change the version of package A in my secondary config file with something like:

```
{
  "require": {
    "A": "dev-feature as dev-trunk"
  }
}
```

I would expect this to work fine because the alias statement should then mean package B's need for `A @ dev-trunk` will still be met (via the alias).

However, I get a conflict resolution failure.

The output differs depending on whether I have "replace" mode true or false but the end result is the same failure to resolve.

### PR DESCRIPTION
**Pull request by jiripudil**
fixes #100

### CODE REVIEW

**Comment by reedy** [comment]
You might want to rebase your patch so CI doesn't error :)

**Comment by jiripudil** [comment]
Good point, rebased onto #189. Should I mark this PR as a draft so that it doesn't get merged accidentally?

**Comment by jiripudil** [comment]
Also, on a side note, would you appreciate a PR migrating over to Github Actions? To my dismay, waiting for Travis feels like ages recently.

**Comment by reedy** [comment]
> Also, on a side note, would you appreciate a PR migrating over to Github Actions? To my dismay, waiting for Travis feels like ages recently.

Could do.

I was planning on doing the same thing when I got a little bit more time this weekend, as it's not particularly difficult to do :). 

Fixing CI (forcing composer 1.x) as I did was only a couple of minute job, so was easily done in the meantime.

**Comment by reedy** [comment]
> Also, on a side note, would you appreciate a PR migrating over to Github Actions? To my dismay, waiting for Travis feels like ages recently.

Started on https://github.com/wikimedia/composer-merge-plugin/tree/githubactions

**Comment by jiripudil** [comment]
Lovely, thank you! Rebased onto master. The development is suddenly a breeze when you get almost instant feedback from the CI :)

**Comment by XedinUnknown** [comment]
Hi! Since this seems to have been fixed some time ago, and tests are passing, what is blocking the merge please? I would like to use Composer's resolution for my modules if they declare the same dep as root package. In my case, the module needs to depend on a fix branch of the same dep that is required with a more stable constraint in the root. If the module was a dependency, this would be trivial to solve with inline aliasing. Alas, it does not work. I'm really hoping this PR could fix this.

**Comment by sajphon** [comment]
Tested the PR locally and it seems to do exactly what it claims to do and I didnt run into any issues while using it. This should be merged asap.

**Comment by XedinUnknown** [comment]
This [comment][1] seems to imply that the changes of this PR are included in the `master` branch. If it has, I guess it has been released? What's the version with this change, please? Since it's a rebase and not a merge, this information is not available. 

[1]: https://github.com/wikimedia/composer-merge-plugin/pull/196#issuecomment-757687111

**Comment by reedy** [comment]
> This [comment](https://github.com/wikimedia/composer-merge-plugin/pull/196#issuecomment-757687111) seems to imply that the changes of this PR are included in the `master` branch. If it has, I guess it has been released? What's the version with this change, please? Since it's a rebase and not a merge, this information is not available.

No it doesn't. It was originally against master before we had GitHub actions. It was rebased onto newer master when we had GitHub actions, so CI actually worked properly.

**Comment by sajphon** [comment]
@reedy so what seems to be the problem with releasing it ? 🤔

**Comment by XedinUnknown** [comment]
Any news on this, please?

**Comment by sajphon** [comment]
@reedy any news ?

**Review by sajphon — APPROVED** [review]

**Comment by utrotzek** [comment]
I stumbled up on that issue today as well. It would be really nice if the fix can be releases in one of the next versions. I can help with testing if desired ;)

**Comment by pradtke** [comment]
Since `jiripudil:preserve-aliases` is a few commits behind `master`, I made my own branch from `wikimedia:master` and merged the PR into that, and tested it. It worked as I expected.  I'm looking forward to this PR getting merged and released.

**Comment by steffendietz** [comment]
Stumbled across #100 basically 5 minutes after discovering this plugin and giving it a spin.

Could this be moved forward, please? It's been more than 6 years the issue was first described and now almost 2 years this seemingly functioning PR is just sitting here.

**Comment by HaaseIT** [comment]
@reedy any news on this?

File: src/ExtraPackage.php [hunk 1]
```diff
@@ -644,6 +646,51 @@ public static function unwrapIfNeeded(
...
+
+    /**
+     * Extract aliases from version constraints (dev-branch as 1.0.0).
+     *
+     * @param array $requires
+     * @param array $aliases
+     * @return array
+     * @see RootPackageLoader::extractAliases()
+     */
+    protected function extractAliases(array $requires, array $aliases)
```
Comment (reedy) [inline src/ExtraPackage.php:673]: Annoying that we can't just use the `RootPackageLoader::extractAliases()` instead of having to duplicate it

**Comment by jiripudil** [comment]
Wow, glad to see this merged, eventually. Thanks! 🫶

**Pull request merged**